data-prepper-core/src/test/java/org/opensearch/dataprepper/peerforwarder/PeerForwarderTest.java

```diff
@@ -0,0 +1,113 @@
+        return new PeerForwarder(peerForwarderClientFactory, peerForwarderClient);
+    }
+
+    @Test
+    void test_forwardRecords_with_two_local_ips_should_process_record_two_record_locally() {
+        when(peerForwarderClientFactory.createHashRing()).thenReturn(hashRing);
+        final List<String> testIps = List.of("127.0.0.1", "128.0.0.1");
+        lenient().when(hashRing.getServerIp(List.of("value1", "value1"))).thenReturn(Optional.of(testIps.get(0)));
+        lenient().when(hashRing.getServerIp(List.of("value2", "value2"))).thenReturn(Optional.of(testIps.get(1)));
+
+        PeerForwarder peerForwarder = createObjectUnderTest();
+
+        final List<Record<Event>> records = peerForwarder.forwardRecords(generateBatchRecords(2, false),
+                generateIdentificationKeys(), TEST_PLUGIN_ID);
+        verifyNoInteractions(peerForwarderClient);
+        assertThat(records.size(), equalTo(2));
+    }
+
+    @Test
+    void test_forwardRecords_with_one_local_ip_and_one_remote_ip_should_process_record_one_record_locally() {
+        AggregatedHttpResponse aggregatedHttpResponse = mock(AggregatedHttpResponse.class);
+        when(peerForwarderClientFactory.createHashRing()).thenReturn(hashRing);
+        when(aggregatedHttpResponse.status()).thenReturn(HttpStatus.OK);
+        when(peerForwarderClient.serializeRecordsAndSendHttpRequest(anyCollection(), anyString(), anyString())).thenReturn(aggregatedHttpResponse);
+
+        final List<String> testIps = List.of("8.8.8.8", "127.0.0.1");
+        lenient().when(hashRing.getServerIp(List.of("value1", "value1"))).thenReturn(Optional.of(testIps.get(0)));
+        lenient().when(hashRing.getServerIp(List.of("value2", "value2"))).thenReturn(Optional.of(testIps.get(1)));
+
+        PeerForwarder peerForwarder = createObjectUnderTest();
+
+        final List<Record<Event>> records = peerForwarder.forwardRecords(generateBatchRecords(2, false),
+                generateIdentificationKeys(), TEST_PLUGIN_ID);
+        verify(peerForwarderClient, times(1)).serializeRecordsAndSendHttpRequest(anyList(), anyString(), anyString());
+        assertThat(records.size(), equalTo(1));
+    }
+
+    private Collection<Record<Event>> generateBatchRecords(final int numRecords, final boolean isSameValues) {
+        final Collection<Record<Event>> results = new ArrayList<>();
+        for (int i = 0; i < numRecords; i++) {
+            final Map<String, String> eventData = new HashMap<>();
+            if (isSameValues) {
+                eventData.put("key1", "value");
+                eventData.put("key2", "value");
+            }
+            else {
+                eventData.put("key1", "value" + i);
+                eventData.put("key2", "value" + i);
+            }
+            final JacksonEvent event = JacksonLog.builder().withData(eventData).build();
+            results.add(new Record<>(event));
+        }
+        return results;
+    }
+
+    private Set<String> generateIdentificationKeys() {
```

Comment:
Our unit tests just scratch the surface. Let's add some more and try to break our code. 

What if keys are empty? What if the peer forwarder client, factory or hashRing throw an exception, return a bad response or failure?